command/agent/retry_join.go

```diff
@@ -30,6 +35,57 @@ type DiscoverInterface interface {
 	Names() []string
 }
 
+// discoverProxy proxies retryJoiner discovery requests to either go-netaddrs or go-discover
+//
+// Addresses are first discovered using go-netaddrs, but if no addresses are able to be disocovered,
+// fall back to go-autodiscover
+type discoverProxy struct {
+	goDiscover *discover.Discover
+}
+
+// IPAddrs looks up and returns IP addresses using the method described by cfg.
+//
+// If cfg is a DNS name, IP addresses are looked up by querying the default
+// DNS resolver for A and AAAA records associated with the DNS name.
+//
+// If cfg has an exec= prefix, IP addresses are looked up by executing the command
+// after exec=. The command may include optional arguments. Command arguments
+// must be space separated (spaces in argument values can not be escaped).
+// The command may output IPv4 or IPv6 addresses, and IPv6 addresses can
+// optionally include a zone index.
+// The executable must follow these rules:
+//
+//	on success - exit 0 and print whitespace delimited IP addresses to stdout.
+//	on failure - exits with a non-zero code, and should print an error message
+//	             of up to 1024 bytes to stderr.
+//
+// if addresses cannot be looked up using DNS or an exec command, fall back to
+// discovering addresses with go-discover
+func (d discoverProxy) Addrs(cfg string, logger *golog.Logger) (addrs []string, err error) {
+	var ipAddrs []net.IPAddr
+	ipAddrs, err = netaddrs.IPAddrs(context.Background(),
+		cfg,
+		log.FromStandardLogger(logger, &log.LoggerOptions{Name: "joiner"}))
+	if err != nil {
+		logger.Print("falling back to go-discover for auto-join config: ", cfg, "due to:", err)
```

Comment:
Most cloud auto-join configs will be for `go-discover` for quite some time. We probably won't want to log to standard out here since this log might imply that something is wrong, even though this is expected behavior. At the very least, we probably don't want to log the error here. 

Alternatively, maybe `Addrs` should take a logger with log levels so this can use a `Debug()` log instead.